container/podman/handler.go

```diff
@@ -0,0 +1,308 @@
+		metrics:            metrics,
+		thinPoolName:       thinPoolName,
+		zfsParent:          zfsParent,
+		reference: info.ContainerReference{
+			Id:        id,
+			Name:      name,
+			Aliases:   []string{strings.TrimPrefix(ctnr.Name, "/"), id},
+			Namespace: Namespace,
+		},
+		libcontainerHandler: containerlibcontainer.NewHandler(cgroupManager, rootFs, ctnr.State.Pid, metrics),
+	}
+
+	handler.creationTime, err = time.Parse(time.RFC3339, ctnr.Created)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the create timestamp %q for container %q: %v", ctnr.Created, id, err)
+	}
+
+	if ctnr.RestartCount > 0 {
+		handler.labels["restartcount"] = fmt.Sprint(ctnr.RestartCount)
+	}
+
+	// Obtain the IP address for the container.
+	// If the NetworkMode starts with 'container:' then we need to use the IP address of the container specified.
+	// This happens in cases such as kubernetes where the containers doesn't have an IP address itself and we need to use the pod's address
+	networkMode := string(handler.networkMode)
+	if handler.ipAddress == "" && strings.HasPrefix(networkMode, "container:") {
+		id := strings.TrimPrefix(networkMode, "container:")
+		ctnr, err := InspectContainer(id)
+		if err != nil {
+			return nil, err
+		}
+		handler.ipAddress = ctnr.NetworkSettings.IPAddress
+	}
+
+	if metrics.Has(container.DiskUsageMetrics) {
+		handler.fsHandler = &docker.FsHandler{
+			FsHandler:       common.NewFsHandler(common.DefaultPeriod, rootfsStorageDir, otherStorageDir, fsInfo),
+			ThinPoolWatcher: thinPoolWatcher,
+			ZfsWatcher:      zfsWatcher,
+			DeviceID:        ctnr.GraphDriver.Data["DeviceId"],
+			ZfsFilesystem:   zfsFilesystem,
+		}
+	}
+
+	// Split env vars to get metadata map.
+	for _, exposedEnv := range metadataEnvAllowList {
+		if exposedEnv == "" {
+			continue
+		}
+
+		for _, envVar := range ctnr.Config.Env {
+			if envVar != "" {
+				splits := strings.SplitN(envVar, "=", 2)
+				if len(splits) == 2 && strings.HasPrefix(splits[0], exposedEnv) {
+					handler.envs[strings.ToLower(splits[0])] = splits[1]
+				}
+			}
+		}
+	}
+
+	return handler, nil
+}
+
+func (p podmanContainerHandler) ContainerReference() (info.ContainerReference, error) {
+	return p.reference, nil
+}
+
+func (p podmanContainerHandler) needNet() bool {
+	if p.metrics.Has(container.NetworkUsageMetrics) {
+		p.networkMode.IsContainer()
+		return !p.networkMode.IsContainer()
+	}
+	return false
+}
+
+func (p podmanContainerHandler) GetSpec() (info.ContainerSpec, error) {
+	hasFilesystem := p.metrics.Has(container.DiskUsageMetrics)
+
+	spec, err := common.GetSpec(p.cgroupPaths, p.machineInfoFactory, p.needNet(), hasFilesystem)
+	if err != nil {
+		return info.ContainerSpec{}, err
+	}
+
+	spec.Labels = p.labels
+	spec.Envs = p.envs
+	spec.Image = p.image
+	spec.CreationTime = p.creationTime
+
+	return spec, nil
+}
+
+func (p podmanContainerHandler) GetStats() (*info.ContainerStats, error) {
```

Comment:
Why not just Stats()

Reply:
`podmanContainerHandler` implements `ContainerHandler` interface. I agree that it should be refactored but maybe in a different PR because the current one is huge.